src/completion.lisp

```diff
@@ -39,10 +41,47 @@
                           :always (funcall test p1 p2)))))
            (test-without-separator (elt)
              (funcall test name (apply-key elt))))
-    (remove-if-not (if separator
-                       #'test-with-separator
-                       #'test-without-separator)
-                   elements)))
+    (let ((filtered-elements
+            (remove-if-not (if separator
+                               #'test-with-separator
+                               #'test-without-separator)
+                           elements)))
+      (if rank
+          (sort filtered-elements #'< :key (lambda (elt) (funcall rank name (apply-key elt))))
+          filtered-elements))))
+
+(defun string-completion-rank (name elt)
+  (cond
+    ; Exact match
+    ((string= name elt) 0)
+    ; Prefix match
+    ((str:starts-with-p name elt) (length name))
+    ; Substring match anywhere
+    ((search name elt) 2)
+    ; Fuzzy match, rank by length
+    (t (length elt))))
+
+(defun file-completion-rank (name elt)
+  (let ((file-name (file-namestring elt)))
+    (cond
+      ; Exact match
+      ((string= name elt) 0)
+      ; Prefix match in file name
+      ((str:starts-with-p name file-name) 1)
+      ; Substring match in file name
+      ((search name file-name) 2)
+      ; Prefix match in full path
+      ((str:starts-with-p name elt) 3)
+      ; Substring match in full path
+      ((search name elt) 4)
+      ; Fuzzy match, rank by length
+      (t (length elt)))))
+
+(defun completion-strings (str strings &key key files)
```

Comment:
I was a little confused because the argument files looks like it passes a list of files.
For future extensibility, could we use a different interface instead of boolean?
For example, for files, it could be completion-files instead of completion-strings.

Reply:
Sure, I wanted to do that at first but there was another `completion-file` down there and I thought it'd be confusing. You think if I move it to a new `completion-files` function that's fine?

Reply:
I see, it might be confusing to have two `completion-files` and `completion-file`.
I thought it would be good to have a `rank` parameter for completion-strings.
This could be refactored in the future.